ansible_base/rbac/triggers.py

```diff
@@ -0,0 +1,336 @@
+    update_fields = kwargs.get('update_fields', None)
+    if update_fields and not (parent_field_name in update_fields or f'{parent_field_name}_id' in update_fields):
+        return
+
+    # If we HAVE to do a query to find out if the parent field has changed then we will here
+    if not hasattr(instance, '__rbac_original_parent_id'):
+        instance.__rbac_original_parent_id = getattr(type(instance).objects.only('id').get(pk=instance.pk), f'{parent_field_name}_id')
+
+
+def rbac_post_save_update_evaluations(instance, created, *args, **kwargs):
+    """
+    Connect to post_save signal for objects in the permission registry
+    If the parent object changes, this rebuilds the cache
+    """
+    # Exit right away if object does not have any parent objects
+    parent_field_name = permission_registry.get_parent_fd_name(instance)
+    if parent_field_name is None:
+        return
+
+    # If child object is created and parent object has existing ObjectRoles
+    # evaluations for the parent object roles need to be added
+    if created:
+        post_save_update_obj_permissions(instance)
+        return
+
+    # The parent object can not have changed if update_fields was given and did not list that field
+    update_fields = kwargs.get('update_fields', None)
+    if update_fields and not (parent_field_name in update_fields or f'{parent_field_name}_id' in update_fields):
+        return
+
+    # Handle the unusual situation where the parent object changes
+    current_parent_id = getattr(instance, f'{parent_field_name}_id')
+    if hasattr(instance, '__rbac_original_parent_id') and instance.__rbac_original_parent_id != current_parent_id:
+        logger.info(f'Object {instance} changed RBAC parent {instance.__rbac_original_parent_id}-->{current_parent_id}')
+        post_save_update_obj_permissions(instance)
+
+
+def team_pre_delete(instance, *args, **kwargs):
+    instance.__rbac_stashed_member_roles = list(instance.member_roles.all())
+
+
+def rbac_post_delete_remove_object_roles(instance, *args, **kwargs):
+    """
+    Call this when deleting an object to cascade delete its object roles
+    Deleting a team can have consequences for the rest of the graph
+    """
+    if instance._meta.model_name == permission_registry.team_model._meta.model_name:
+        indirectly_affected_roles = set()
+        indirectly_affected_roles.update(team_ancestor_roles(instance))
+        for team_role in instance.__rbac_stashed_member_roles:
+            indirectly_affected_roles.update(team_role.descendent_roles())
+        compute_team_member_roles()
+        compute_object_role_permissions(object_roles=indirectly_affected_roles)
+
+        # Similar to user deletion, clean up any orphaned object roles
+        ObjectRole.objects.filter(users__isnull=True, teams__isnull=True).delete()
+
+    ct = ContentType.objects.get_for_model(instance)
+    ObjectRole.objects.filter(content_type=ct, object_id=instance.id).delete()
```

Comment:
in the case where I have Org > Inventory, and I delete an Inventory, I notice the RoleEvaluations associated with that Inventory (at the Org level) are not deleted right away. Not until `compute_object_role_permissions` is called again.

Trying to think if having stale RoleEvaluation entries could lead to any issues. I guess since the Inventory no longer exists, these RoleEvaluations are benign

Reply:
This test case passes for me:

```python
@pytest.mark.django_db
def test_seth(org_inv_rd, rando, organization, inventory):
    org_inv_rd.give_permission(rando, organization)
    inv_ct = permission_registry.content_type_model.objects.get_for_model(inventory)
    assert RoleEvaluation.objects.filter(object_id=inventory.id, content_type_id=inv_ct.id).count() == 2
    inventory.delete()
    assert RoleEvaluation.objects.filter(object_id=inventory.id, content_type_id=inv_ct.id).count() == 0
```

Now if you're talking about the case where `ANSIBLE_BASE_CACHE_PARENT_PERMISSIONS` is set to True, it will make `RoleEvaluation` entries corresponding to the `organization` object too. It should still do that even if there are no inventories in the organization, and rebuilding the cache shouldn't remove those entries either.

Reply:
that test makes sense, here is what I was seeing in shell_plus
```
In [21]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[21]: 24

In [22]: i = Inventory.objects.create(name="invA", organization=o)
2024-03-05 18:33:54,836 DEBUG    ansible_base.rbac.caching Adding 4 object-permissions to ObjectRole(pk=1, organization=1)
2024-03-05 18:33:54,836 INFO     ansible_base.rbac.caching Adding 4 object-permission records

In [23]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[23]: 28
```

so far so good, we added an inventory to an org, which gets 4 more RoleEvaluation entries, one for each permission

```
In [24]: i.delete()
Out[24]: (1, {'test_app.Inventory': 1})

In [25]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[25]: 28
```

the count didn't go down as I expected

```
In [26]: i = Inventory.objects.create(name="invA", organization=o)
2024-03-05 18:34:12,372 DEBUG    ansible_base.rbac.caching Removing 4 object-permissions from ObjectRole(pk=1, organization=1)
2024-03-05 18:34:12,372 DEBUG    ansible_base.rbac.caching Adding 4 object-permissions to ObjectRole(pk=1, organization=1)
2024-03-05 18:34:12,372 INFO     ansible_base.rbac.caching Adding 4 object-permission records
2024-03-05 18:34:12,390 INFO     ansible_base.rbac.caching Deleting 4 object-permission records

In [27]: RoleEvaluation.objects.filter(content_type_id=31).count()
Out[27]: 28

```
adding a new inventory does trigger some log lines that say 4 were removed, and 4 were added

Reply:
Gah! Because `inventory.id` was referenced after it was deleted in my test case. Now I'm reproducing, irrelevant of that setting. Yikes, I'm going to have to spend some time chasing this down and buttoning it up. Nice find!